File: simulator/python/board/profiles.py
"""Perfis de boards para o virtual board, espelhando `board-generic/src/profiles.rs`.

Valores (name, aaguid, transports, secure storage, crypto accelerator e pinos)
são mantidos em paralelo com os perfis Rust de `firmware/board-generic`:
NRF52840, STM32L4, ESP32C3, RP2350 e GENERIC.
"""

from dataclasses import dataclass, field

TRANSPORT_USB_CCID = 0x01
TRANSPORT_USB_HID = 0x02
TRANSPORT_NFC = 0x04
TRANSPORT_BLE = 0x08


@dataclass(frozen=True)
class SecurityFeatures:
    secure_boot: bool = False
    trust_zone: bool = False
    hardware_rng: bool = False
    sha256_accelerator: bool = False
    debug_disable: bool = False
    otp_memory: bool = False
    unique_id: bool = False
    tamper_detection: bool = False


@dataclass(frozen=True)
class BoardProfile:
    name: str
    aaguid: bytes
    transports: int
    has_secure_storage: bool
    has_crypto_accelerator: bool
    security: SecurityFeatures = None
    i2c_sda_pin: int = 0
    i2c_scl_pin: int = 0
    spi_mosi_pin: int = 0
    spi_miso_pin: int = 0
    spi_clk_pin: int = 0
    cs_pin: int = 0
    reset_pin: int = 0
    irq_pin: int = 0
    led_pin: int = 0
    button_pin: int = 0

    def has_transport(self, mask):
        return bool(self.transports & mask)


NRF52840 = BoardProfile(
    name="nrf52840-fido",
    aaguid=bytes([0x4E, 0x52, 0x46, 0x35, 0x32, 0x38, 0x34, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01]),
    transports=TRANSPORT_USB_HID | TRANSPORT_NFC | TRANSPORT_BLE,
    has_secure_storage=True,
    has_crypto_accelerator=True,
    i2c_sda_pin=4, i2c_scl_pin=5,
    spi_mosi_pin=6, spi_miso_pin=7, spi_clk_pin=8, cs_pin=9,
    reset_pin=10, irq_pin=11, led_pin=12, button_pin=13,
)

STM32L4 = BoardProfile(
    name="stm32l4-fido",
    aaguid=bytes([0x53, 0x54, 0x4D, 0x33, 0x32, 0x4C, 0x34, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02]),
    transports=TRANSPORT_USB_HID | TRANSPORT_USB_CCID,
    has_secure_storage=True,
    has_crypto_accelerator=True,
    led_pin=0, button_pin=1,
    i2c_sda_pin=8, i2c_scl_pin=9,
    spi_mosi_pin=10, spi_miso_pin=11, spi_clk_pin=12, cs_pin=13,
    reset_pin=14, irq_pin=15,
)

ESP32C3 = BoardProfile(
    name="esp32c3-fido",
    aaguid=bytes([0x45, 0x53, 0x50, 0x33, 0x32, 0x43, 0x33, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03]),
    transports=TRANSPORT_USB_HID | TRANSPORT_BLE,
    has_secure_storage=False,
    has_crypto_accelerator=True,
    i2c_sda_pin=4, i2c_scl_pin=5,
    spi_mosi_pin=6, spi_miso_pin=7, spi_clk_pin=8, cs_pin=9,
    reset_pin=10, irq_pin=11, led_pin=12, button_pin=13,
)

RP2350 = BoardProfile(
    name="rp2350-fido",
    aaguid=bytes([0x52, 0x50, 0x32, 0x33, 0x35, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x05]),
    transports=TRANSPORT_USB_HID | TRANSPORT_USB_CCID,
    has_secure_storage=True,
    has_crypto_accelerator=True,
    security=SecurityFeatures(
        secure_boot=True,
        trust_zone=True,
        hardware_rng=True,
        sha256_accelerator=True,
        debug_disable=True,
        otp_memory=True,
        unique_id=True,
        tamper_detection=False,
    ),
    i2c_sda_pin=4, i2c_scl_pin=5,
    spi_mosi_pin=6, spi_miso_pin=7, spi_clk_pin=8, cs_pin=9,
    reset_pin=10, irq_pin=11, led_pin=25, button_pin=13,
)

GENERIC = BoardProfile(
    name="generic-fido",
    aaguid=bytes([0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF]),
    transports=TRANSPORT_USB_CCID,
    has_secure_storage=False,
    has_crypto_accelerator=False,
    i2c_sda_pin=0, i2c_scl_pin=1,
    spi_mosi_pin=2, spi_miso_pin=3, spi_clk_pin=4, cs_pin=5,
    reset_pin=6, irq_pin=7, led_pin=8, button_pin=9,
)

ALL = [NRF52840, STM32L4, ESP32C3, RP2350, GENERIC]

File: firmware/board-generic/src/profiles.rs
//! Pre-defined board profiles built on [`BoardDefinition`].
//!
//! Each profile is a `const` value describing a target board's identity
//! (name and AAGUID), the physical transports it exposes, the hardware
//! features it provides (secure storage, crypto accelerator) and its logical
//! HAL pin assignments. Profiles can be passed directly to
//! `DeviceProfileBuilder::from_board` or `EmbeddedAuthenticator::new_with_board`.
//!
//! Available profiles: `NRF52840`, `STM32L4`, `ESP32C3`, `RP2350` and `GENERIC`.

use crate::board_generic::{BoardDefinition, SecurityFeatures};

/// nRF52840-based authenticator with USB-HID, NFC and BLE transports.
pub const NRF52840: BoardDefinition = BoardDefinition::new(
    "nrf52840-fido",
    [
        0x4e, 0x52, 0x46, 0x35, 0x32, 0x38, 0x34, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x01,
    ],
)
.usb_hid()
.nfc()
.ble()
.secure_storage(true)
.crypto_accelerator(true)
.i2c_sda(4)
.i2c_scl(5)
.spi_mosi(6)
.spi_miso(7)
.spi_clk(8)
.cs(9)
.reset(10)
.irq(11)
.led(12)
.button(13);

/// STM32L4-based authenticator with USB-HID and USB-CCID transports.
pub const STM32L4: BoardDefinition = BoardDefinition::new(
    "stm32l4-fido",
    [
        0x53, 0x54, 0x4d, 0x33, 0x32, 0x4c, 0x34, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x02,
    ],
)
.usb_hid()
.usb_ccid()
.secure_storage(true)
.crypto_accelerator(true)
.led(0)
.button(1)
.i2c_sda(8)
.i2c_scl(9)
.spi_mosi(10)
.spi_miso(11)
.spi_clk(12)
.cs(13)
.reset(14)
.irq(15);

/// ESP32-C3-based authenticator with USB-HID and BLE transports.
pub const ESP32C3: BoardDefinition = BoardDefinition::new(
    "esp32c3-fido",
    [
        0x45, 0x53, 0x50, 0x33, 0x32, 0x43, 0x33, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x03,
    ],
)
.usb_hid()
.ble()
.crypto_accelerator(true)
.secure_storage(false)
.i2c_sda(4)
.i2c_scl(5)
.spi_mosi(6)
.spi_miso(7)
.spi_clk(8)
.cs(9)
.reset(10)
.irq(11)
.led(12)
.button(13);

/// Raspberry Pi RP2350-based authenticator with USB-HID and USB-CCID transports.
///
/// The RP2350 provides hardware security features including ARM TrustZone
/// (Cortex-M33), secure boot with RSA signature verification, hardware TRNG,
/// SHA-256 accelerator, OTP memory for key storage, unique chip ID, and
/// debug disable via OTP.
///
/// Pin assignments reflect a typical dev board. Use [`rp2350_with_pins`] to
/// override for boards with different wiring.
pub const RP2350: BoardDefinition = rp2350_with_pins(Rp2350Pins {
    i2c_sda: 4,
    i2c_scl: 5,
    spi_mosi: 6,
    spi_miso: 7,
    spi_clk: 8,
    cs: 9,
    reset: 10,
    irq: 11,
    led: 25,
    button: 13,
});

/// RP2350 board with custom pin assignments.
///
/// Use this when the target board uses different GPIOs than the defaults.
///
/// ```
/// use board_generic::profiles::{rp2350_with_pins, Rp2350Pins};
///
/// let board = rp2350_with_pins(Rp2350Pins {
///     i2c_sda: 2,
///     i2c_scl: 3,
///     spi_mosi: 16,
///     spi_miso: 19,
///     spi_clk: 18,
///     cs: 17,
///     reset: 20,
///     irq: 21,
///     led: 10,
///     button: 15,
/// });
/// ```
#[inline]
pub const fn rp2350_with_pins(pins: Rp2350Pins) -> BoardDefinition {
    BoardDefinition::new(
        "rp2350-fido",
        [
            0x52, 0x50, 0x32, 0x33, 0x35, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
            0x00, 0x05,
        ],
    )
    .usb_hid()
    .usb_ccid()
    .secure_storage(true)
    .crypto_accelerator(true)
    .security_features(SecurityFeatures::rp2350())
    .i2c_sda(pins.i2c_sda)
    .i2c_scl(pins.i2c_scl)
    .spi_mosi(pins.spi_mosi)
    .spi_miso(pins.spi_miso)
    .spi_clk(pins.spi_clk)
    .cs(pins.cs)
    .reset(pins.reset)
    .irq(pins.irq)
    .led(pins.led)
    .button(pins.button)
}

/// Pin assignments for [`rp2350_with_pins`].
pub struct Rp2350Pins {
    /// GPIO do sinal I2C SDA.
    pub i2c_sda: u8,
    /// GPIO do sinal I2C SCL.
    pub i2c_scl: u8,
    /// GPIO do sinal SPI MOSI.
    pub spi_mosi: u8,
    /// GPIO do sinal SPI MISO.
    pub spi_miso: u8,
    /// GPIO do clock SPI.
    pub spi_clk: u8,
    /// GPIO do chip select.
    pub cs: u8,
    /// GPIO de reset do periférico.
    pub reset: u8,
    /// GPIO de interrupção.
    pub irq: u8,
    /// GPIO do LED de status (user presence).
    pub led: u8,
    /// GPIO do botão de user presence.
    pub button: u8,
}

/// Generic board profile with USB-CCID transport only.
pub const GENERIC: BoardDefinition = BoardDefinition::new(
    "generic-fido",
    [
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0xff,
    ],
)
.usb_ccid()
.i2c_sda(0)
.i2c_scl(1)
.spi_mosi(2)
.spi_miso(3)
.spi_clk(4)
.cs(5)
.reset(6)
.irq(7)
.led(8)
.button(9);
